src/credentials/GroupcastDataProvider.cpp

```diff
@@ -0,0 +1,245 @@
+    }
+
+    void ClearEntry(Group & entry) override
+    {
+        entry.group_id       = 0;
+        entry.endpoint_count = 0;
+        memset(entry.endpoints, 0x00, sizeof(entry.endpoints));
+    }
+
+    CHIP_ERROR Serialize(TLV::TLVWriter & writer, const Group & entry) const override
+    {
+        ReturnErrorOnFailure(writer.Put(Tags::GroupId(), static_cast<uint16_t>(entry.group_id)));
+        ReturnErrorOnFailure(writer.Put(Tags::KeyId(), static_cast<uint32_t>(entry.key_id)));
+        ReturnErrorOnFailure(writer.Put(Tags::EndpointCount(), static_cast<uint16_t>(entry.endpoint_count)));
+        // Endpoints
+        {
+            TLV::TLVType array, item;
+            ReturnErrorOnFailure(writer.StartContainer(Tags::EndpointList(), TLV::kTLVType_Array, array));
+            for (size_t i = 0; i < entry.endpoint_count; ++i)
+            {
+                ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, item));
+                // Endpoint
+                ReturnErrorOnFailure(writer.Put(Tags::EndpointEntry(), entry.endpoints[i]));
+                ReturnErrorOnFailure(writer.EndContainer(item));
+            }
+            ReturnErrorOnFailure(writer.EndContainer(array));
+        }
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Deserialize(TLV::TLVReader & reader, Group & entry) override
+    {
+        // Group Id
+        ReturnErrorOnFailure(reader.Next(Tags::GroupId()));
+        ReturnErrorOnFailure(reader.Get(entry.group_id));
+        // Key Id
+        ReturnErrorOnFailure(reader.Next(Tags::KeyId()));
+        ReturnErrorOnFailure(reader.Get(entry.key_id));
+        // Endpoint Count
+        ReturnErrorOnFailure(reader.Next(Tags::EndpointCount()));
+        ReturnErrorOnFailure(reader.Get(entry.endpoint_count));
+        // Endpoints
+        ReturnErrorOnFailure(reader.Next(Tags::EndpointList()));
+        VerifyOrReturnError(TLV::kTLVType_Array == reader.GetType(), CHIP_ERROR_INTERNAL);
+        {
+            TLV::TLVType array, item;
+            ReturnErrorOnFailure(reader.EnterContainer(array));
+            for (size_t i = 0; i < entry.endpoint_count; ++i)
+            {
+                ReturnErrorOnFailure(reader.Next(TLV::AnonymousTag()));
+                VerifyOrReturnError(TLV::kTLVType_Structure == reader.GetType(), CHIP_ERROR_INTERNAL);
+                ReturnErrorOnFailure(reader.EnterContainer(item));
+                // Endpoint
+                ReturnErrorOnFailure(reader.Next(Tags::EndpointEntry()));
+                ReturnErrorOnFailure(reader.Get(entry.endpoints[i]));
+                ReturnErrorOnFailure(reader.ExitContainer(item));
+            }
+            ReturnErrorOnFailure(reader.ExitContainer(array));
+        }
+        return CHIP_NO_ERROR;
+    }
+
+    FabricIndex mFabric;
+};
+
+//
+// Group Iterator
+//
+
+DataProvider::GroupIterator::GroupIterator(DataProvider & group_data, FabricIndex fabric) : mProvider(group_data), mFabric(fabric)
```

Comment:
Same here, if we could leverage the TableEntry type, we could use the TableEntryDataConvertingIterator